File: packages/renderer/src/ai.ts
import type { ARLSEvent } from '@agentlens/core'

interface ClaudeContext {
  summary: string
  debug_suggestion?: string
  related_steps?: number[]
  cost_so_far_usd?: number
}

/** AI-mode specific event with Claude context */
interface AIEvent extends ARLSEvent {
  _claude_context: ClaudeContext
}

/** Generate a summary for the Claude context based on event type */
function generateSummary(event: ARLSEvent): string {
  switch (event.schema_type) {
    case 'AGENT_START':
      return `Agent '${event.agent.name}' started execution (run: ${event.run_id})`

    case 'AGENT_END':
      return `Agent '${event.agent.name}' completed execution after ${event.step_index} steps`

    case 'TOOL_CALL': {
      if (!event.tool) return 'Tool call event'
      const duration = (event.tool.duration_ms / 1000).toFixed(2)
      return `Tool '${event.tool.name}' ${event.tool.status.toLowerCase()} in ${duration}s`
    }

    case 'LLM_CALL': {
      if (!event.llm) return 'LLM call event'
      return `LLM call to ${event.llm.model} used ${event.llm.total_tokens} tokens (${event.llm.prompt_tokens} in / ${event.llm.completion_tokens} out)`
    }

    case 'ERROR': {
      if (!event.error) return 'Error event'
      return `Error [${event.error.code}]: ${event.error.message}`
    }

    case 'REASONING_STEP':
      return `Reasoning step by ${event.agent.name} in phase ${event.agent.phase}`

    case 'MEMORY_READ':
      return `Memory read operation (context usage: ${event.memory?.context_window_used_pct ?? 0}%)`

    case 'MEMORY_WRITE':
      return `Memory write operation (context usage: ${event.memory?.context_window_used_pct ?? 0}%)`

    case 'COST_CHECKPOINT': {
      const cost = event.metadata?.total_cost_usd
      return `Cost checkpoint: total spend ${cost ? `$${Number(cost).toFixed(4)}` : 'unknown'}`
    }

    default:
      return `${event.schema_type} event at step ${event.step_index}`
  }
}

/** Generate debug suggestions for common patterns */
function generateDebugSuggestion(event: ARLSEvent): string | undefined {
  if (event.schema_type === 'TOOL_CALL' && event.tool) {
    if (event.tool.status === 'FAILURE') {
      return `Tool failed with: ${event.tool.error_message || 'unknown error'}. Consider retrying with different input or adding error handling.`
    }
    if (event.tool.duration_ms > 5000) {
      return `Tool took ${(event.tool.duration_ms / 1000).toFixed(2)}s to complete. Consider implementing caching or optimizing the tool implementation.`
    }
  }

  if (event.schema_type === 'ERROR') {
    if (event.error?.code === 'CONTEXT_OVERFLOW') {
      return 'Context window exceeded. Consider summarizing earlier messages or implementing context management.'
    }
    if (event.error?.code === 'LOOP_DETECTED') {
      return 'Agent appears to be in a loop. Check decision logic and tool outputs.'
    }
  }

  if (event.ai_debug_hint) {
    return event.ai_debug_hint
  }

  return undefined
}

/**
 * Render a single ARLS event as JSON with Claude-specific metadata for AI tools.
 * Output is JSONL format (one compact JSON object per line, no pretty-printing).
 *
 * @param event - The ARLS event to render
 * @returns Compact JSON string suitable for Claude Code or Copilot
 *
 * @example
 * const event = buildLLMEvent({ ... })
 * const json = renderEventAI(event)
 * console.log(json) // outputs one JSON line
 */
export function renderEventAI(event: ARLSEvent): string {
  const claudeContext: ClaudeContext = {
    summary: generateSummary(event),
  }

  const debugSuggestion = generateDebugSuggestion(event)
  if (debugSuggestion !== undefined) {
    claudeContext.debug_suggestion = debugSuggestion
  }

  const relatedSteps = event.metadata?.related_steps as number[] | undefined
  if (relatedSteps !== undefined) {
    claudeContext.related_steps = relatedSteps
  }

  const costSoFar = event.metadata?.cost_so_far_usd as number | undefined
  if (costSoFar !== undefined) {
    claudeContext.cost_so_far_usd = costSoFar
  }

  const aiEvent: AIEvent = {
    ...event,
    _claude_context: claudeContext,
  }

  // Compact JSON output (no spaces, one line)
  return JSON.stringify(aiEvent)
}

/**
 * Render multiple ARLS events as JSONL output.
 * Each line is a valid JSON object, suitable for streaming or logging.
 *
 * @param events - Array of ARLS events in order
 * @returns JSONL string (one JSON object per line)
 *
 * @example
 * const events = readJSONLFile('agentlens.log')
 * const jsonl = renderSequenceAI(events)
 * console.log(jsonl)
 */
export function renderSequenceAI(events: ARLSEvent[]): string {
  return events.map((event) => renderEventAI(event)).join('\n')
}

File: python/agentlens-renderer/agentlens_renderer/ai.py
"""
AI-readable output renderer for ARLS events.

Produces compact JSONL (one JSON object per line) with additional context
that makes it easy for Claude Code and other AI development tools to
understand what happened and suggest improvements.
"""

import json
from typing import Any

from agentlens_core import ARLSEvent, SchemaType, CallStatus


def render_ai(event: ARLSEvent) -> str:
    """
    Render an ARLS event as JSONL (one compact JSON line).

    Args:
        event: The ARLS event to render

    Returns:
        Single-line JSON string (JSONL format)
    """
    # Start with standard ARLS event dict
    event_dict = event.to_dict()

    # Add Claude-specific context
    claude_context = _generate_claude_context(event)
    if claude_context:
        event_dict['_claude_context'] = claude_context

    # Serialize to JSON (compact, no pretty-printing for JSONL format)
    return json.dumps(event_dict, separators=(',', ':'), ensure_ascii=True)


def _generate_claude_context(event: ARLSEvent) -> dict[str, Any]:
    """
    Generate Claude-specific debug context for an event.

    This helps Claude Code understand the event's significance and
    suggest improvements to agent behavior.

    Args:
        event: The ARLS event

    Returns:
        Dictionary with Claude context fields
    """
    context: dict[str, Any] = {}

    # Generate summary
    if event.schema_type == SchemaType.LLM_CALL and event.llm:
        context['summary'] = (
            f'LLM call to {event.llm.model} '
            f'returned {event.llm.completion_tokens} tokens '
            f'in {event.llm.latency_ms:.0f}ms'
        )
        context['cost_usd'] = event.llm.cost_usd

        # Add performance suggestion
        if event.llm.latency_ms > 5000:
            context['debug_suggestion'] = (
                'LLM call took >5s. Consider caching prompts or using faster models.'
            )
        elif event.llm.completion_tokens > 2000:
            context['debug_suggestion'] = (
                'High output token count. Consider constraining generation.'
            )

    elif event.schema_type == SchemaType.TOOL_CALL and event.tool:
        status_word = (
            'succeeded' if event.tool.status == CallStatus.SUCCESS else 'failed'
        )
        context['summary'] = (
            f'Tool {event.tool.name!r} {status_word} '
            f'in {event.tool.duration_ms:.0f}ms'
        )

        # Error context
        if event.tool.status == CallStatus.FAILURE:
            context['debug_suggestion'] = (
                f'Tool failed with: {event.tool.error_message}. '
                'Consider error handling or fallback logic.'
            )
        elif event.tool.duration_ms > 3000:
            context['debug_suggestion'] = (
                'Tool call was slow. Consider timeout or async optimization.'
            )

    elif event.schema_type == SchemaType.ERROR and event.error:
        context['summary'] = f'{event.error.code}: {event.error.message}'

        # Provide recovery suggestions
        if 'context' in (event.error.message.lower() or ''):
            context['debug_suggestion'] = (
                'Context overflow detected. '
                'Consider summarizing previous messages or using sliding window.'
            )
        elif 'timeout' in (event.error.message.lower() or ''):
            context['debug_suggestion'] = (
                'Operation timed out. '
                'Consider increasing timeout or optimizing the operation.'
            )
        elif event.error.recoverable:
            context['debug_suggestion'] = (
                'Error is recoverable. '
                'Consider implementing retry logic with exponential backoff.'
            )

    elif event.schema_type == SchemaType.AGENT_START:
        context['summary'] = f'Agent {event.agent.name} started execution'

    elif event.schema_type == SchemaType.AGENT_END:
        # Summary of entire run
        total_tokens = event.metadata.get('total_tokens', 0) if event.metadata else 0
        total_cost = event.metadata.get('total_cost_usd', 0) if event.metadata else 0
        step_count = event.metadata.get('step_count', 0) if event.metadata else 0

        context['summary'] = (
            f'Agent {event.agent.name} completed in {step_count} steps'
        )
        context['total_tokens'] = total_tokens
        context['total_cost_usd'] = total_cost

    # Add step reference
    context['step_index'] = event.step_index
    context['trace_id'] = event.trace_id[:12]  # Abbreviated trace ID

    # Add PII status if present
    if event.privacy.pii_detected:
        context['pii_redacted'] = len(event.privacy.redacted_fields) > 0
        context['redacted_field_count'] = len(event.privacy.redacted_fields)

    return context


def render_ai_compact(event: ARLSEvent) -> dict[str, Any]:
    """
    Render an ARLS event as a compact dictionary (not JSON string).

    Useful for programmatic access rather than JSONL output.

    Args:
        event: The ARLS event

    Returns:
        Dictionary representation of the event
    """
    event_dict = event.to_dict()
    claude_context = _generate_claude_context(event)
    if claude_context:
        event_dict['_claude_context'] = claude_context
    return event_dict
